ngraph/core/src/op/assign.cpp

```diff
@@ -60,22 +64,26 @@ void op::v3::Assign::validate_and_infer_types()
     auto variable_info = m_variable->get_info();
     NODE_VALIDATION_CHECK(this,
                           m_variable_id == variable_info.variable_id,
-                          "Variables identifiers are inconsistent.");
-    NODE_VALIDATION_CHECK(
-        this, arg_t == variable_info.data_type, "Variables types are inconsistent.");
+                          "Variables identifiers are inconsistent. Got: ",
+                          m_variable_id,
+                          " and ",
+                          variable_info.variable_id);
 
-    if (output_shape.is_static() && variable_info.data_shape.is_static())
-    {
-        NODE_VALIDATION_CHECK(this,
-                              output_shape == variable_info.data_shape,
-                              "Variables output shapes are inconsistent.");
+    NODE_VALIDATION_CHECK(this,
+                          arg_t == variable_info.data_type,
+                          "Variables types are inconsistent. Got: ",
+                          arg_t,
+                          " and ",
+                          variable_info.data_type);
 
-        set_output_type(0, arg_t, output_shape);
-    }
-    else
-    {
-        set_output_type(0, arg_t, PartialShape::dynamic());
-    }
+    NODE_VALIDATION_CHECK(this,
+                          output_pshape.compatible(variable_info.data_shape),
+                          "Variables output shapes are inconsistent. Got: ",
+                          output_pshape,
+                          " and ",
+                          variable_info.data_shape);
+
+    set_output_type(0, arg_t, variable_info.data_shape);
```

Comment:
Setting `variable_info.data_shape` as output shape is to provide output with the most possible specialized value (the least number of dynamic dimensions as possible).